### ISSUE TITLE
Cadence/Innovus router should FAIL if it created short circuits.

### ISSUE THREAD

**Issue by steveri** [mflowgen/mflowgen#119]
Several times, I have had the experience where a CI flow will reach its final step (LVS) only to fail because an earlier step silently failed routing and left short circuits in the design. This means an extra hour or two of wasted runtime waiting to find out about an error that should have been flagged much earlier.

Why does the router do this? I discovered that, by default, the Cadence/Innovus `routeDesign` command *is* indeed supposed to fail automatically when short circuits result. EXCEPT! According to the Innovus Text Command Reference v. 19.10 p. 3204:

> *Note: -drouteAutoStop is false by default for 20nm designs and below.*

So annoying. Anyway. I can think of several optons for addressing this problem:

1. Ignore it.

2. Add an extra step in the CI pipeline to check for shorts after routing. This would probably involve an ugly hack script in `garnet/.buildkite/bin` that would go into the route step, fire up innovus, and run a verification script.

3. Add an extra mflowgen node to check for shorts after completing the route step. This would involve identifying a step to hold the verification script (either by hijacking an existing step or creating a new step e.g. `custom_postroute_script`); connecting that step to each route node; and doing that for each module with a route step (e.g. `fullchip, Tile_PE, Tile_MemCore, global buffer`...)

4. Modify mflowgen itself to include the check, which should be as simple as modifying the `innovus-flow` step, similar to what happened in mflowgen pull 111 ( https://github.com/mflowgen/mflowgen/pull/111 ). That time, for similar reasons, I changed
```
  - routeDesign
  + routeDesign -placementCheck
```
and this time I would add "-dRouteAutoStop":
```
  - routeDesign -placementCheck
  + setNanoRouteMode -drouteAutoStop true
  + routeDesign -placementCheck
```
Option 4 seems like the correct solution, for so many reasons. So I will proceed with that until/unless I hear a better suggestion.

As usual, your comments are more than welcome!

**Comment by ctorng** [comment]
If we always wanted it to stop if there are shorts, then your number 4 (or an mflowgen assertion) sound like the right way to go.

However, sometimes people do still want to continue the spin despite shorts during prototyping. I think the default should let it run through, since even if detailed route isn't completely clean, the downstream reports are likely to still be useful.

Once node inheritance is up, I think the right thing to do is to sub-class the route node and bake into it the fact that we would like it to stop on shorts for this project.

**Comment by steveri** [comment]
Note that if you are looking for consistent behavior, the "more or less than 20nm" rule referred to above has already thrown that out the window :(

How about this, for a nice low-tech solution, which has the advantage of being wildly generic such that it might solve multiple other little similar problems in the future. In the spirit of generators, we parameterize the route command, i.e. we can change the route step to do use an environment variable / parameter something like
```
if {[info exists ::env(pre_routeDesign)]} {  eval $::env(pre_routeDesign) }
routeDesign -placementCheck
```
Then I can achieve my desired CI behavior by simply setting the parameter
```
export pre_routeDesign="setNanoRouteMode -drouteAutoStop true"
```
Meanwhile, in the case where e.g. the design technology is less than 20nm, the user can optionally turn on the prototyping behavior that you wanted, by setting
```
export pre_detailRoute="setNanoRouteMode -drouteAutoStop false"
```

**Comment by ctorng** [comment]
I like the parameter approach. The parameter should probably be very specific (e.g., "stop_on_failed_routing"). If it is true, then you tack on the additional flag.

**Comment by steveri** [comment]
To be clear @ctorng : you would favor an approach with a simple parameter
```
set stop_on_failed_routing true
```
but would not like the more generic solution
```
set execute_before_detailRoute="setNanoRouteMode -drouteAutoStop true"
```
Correct?

**Comment by ctorng** [comment]
Yes, let's make our parameters specific.

**Comment by steveri** [comment]
I am abandoning this effort. "-drouteAutoStop" did not do what I wanted, in particular it did not produce any kind of error in the case of e.g. metal shorts after routing. I think maybe it only kicks in if there are more than some certain number of errors, which does not appear to be settable AFAICT. Secondly, I found that metal shorts left after initial routing can and do occasionally get fixed successfully in later design stages, so having this "early out" is not as useful as I first anticipated.
